scylla-gdb: active sstables: support k_l/mc sstable readers

bhalevy opened the pull request:
Fixes #5277

Tests: (gdb) scylla active sstables

In v3:
- Defined _lookup_obj helper to simplify cache lookup

In v2:
- Simplified construction of `types` list of dicts
- Added break after yielding result
- Added a cache for resolved vtable_addr

bhalevy commented:
@tgrabiec can you please merge this PR?

tgrabiec commented on scylla-gdb.py:
```diff
@@ -838,18 +838,29 @@ def _lookup_type(type_name):
+            pass
+        # insert negative entry in cache in case
+        # vtable_addr is not found or not matching any reader type
+        vtable_addr_to_ptr_type[vtable_addr] = None
```
I think this can make `vtable_addr_to_ptr_type` grow to O(core dump size), which may overflow host's memory.

Rather than optimizing this, we should do a code change similar to 33bef82f6bb30e6623b84a18a2ac245f3292042a

denesb replied:
I suggested this optimization because it greatly reduced the runtime of another (similar) command that I wrote. However I just found out today, that if there are millions of virtual objects in memory (happens often when scylla dumps core), the command still takes hours to run becoming effectively useless :(.
So I agree with Tomek, commands whose runtime scales with the number of virtual objects will be of limited usefullness at best.

I think we can go one step further and link all readers into an intrusive list. Filtering through that list will be much faster than filtering all virtual objects.

bhalevy replied:
I was under the (false) assumption that vtable_addr is O(#types) rather than O(#objects)...

I'll resend just the first, unoptimized, patch then and later we can better track the readers in scylla as Botond suggested.

I see that we have the same scalability issue in `resolve` with addr -> name mapping and it has a `cache` param to enable the cache. We can add a flag to `scylla active sstables` or even a global flag (e.g. `scylla set caching [on|off]` command) that will control default metadata caching in scylla-gdb so we could turn it off if we run out of memory due to having too many objects.

denesb replied:
`vtable_addr` *is* `O(#types)`. It is the scan of the entire memory that is `O(#objects)`. Caching speeds that up to some extent but if there are millions of objects the scan still takes a lot of time (hours even).

bhalevy replied:
Yes. this pr is not attempting to fix that.
I suggest we merge this one since it does improve functionality until we do the intrusive list of sstable readers, and do the latter as a follow up.

denesb replied:
Agree.

bhalevy commented:
In v4:
- Removed `vtable_addr_to_ptr_type` patch

In v5:
- Added `scylla-gdb: resolve: add startswith parameter`
  to deal with the potential host memory explosion when calling `resolve` on all values returned by `find_vptrs`

bhalevy commented:
@tgrabiec please re-review
